web3/utils/datastructures.py

```diff
@@ -0,0 +1,70 @@
+
+from collections import (
+        Mapping,
+        MutableMapping,
+        Hashable,
+        )
+
+# Hashable must be immutable:
+# "the implementation of hashable collections requires that a key's hash value is immutable"
+# https://docs.python.org/3/reference/datamodel.html#object.__hash__
+
+
+class ReadableAttributeDict(Mapping):
+    'the read attributes for the AttributeDict types'
+
+    def __init__(self, dictionary, *args, **kwargs):
+        self.__dict__ = dict(dictionary)
+        self.__dict__.update(dict(*args, **kwargs))
+
+    def __getitem__(self, key):
+        return self.__dict__[key]
+
+    def __iter__(self):
+        return iter(self.__dict__)
+
+    def __len__(self):
+        return len(self.__dict__)
+
+    def __str__(self):
+        return repr(self)
+
+    def __repr__(self):
+        dict_lines = ['\n\t' + repr(k) + ': ' + repr(self[k]) + ',' for k in sorted(self.__dict__)]
```

Comment:
That was my first implementation until I saw it in the interpreter/docs, which will get uglier. Showing a block/transaction in REPL is a pretty common use case. How about leaving it in for prettier interpreter usage and documentation?

Old docs:
```
>>> web3.eth.getBlock('latest')
{
  'number': 4023147,
  'hash': "0xf3eb83d76f27eddf21ca9db87c297612c07fad00807a601e03990feb5b6c9ad4"
  ... etc ...
}
```

Current docs:
```
>>> web3.eth.getBlock('latest')
AttributeDict({
  'number': 4023147,
  'hash': "0xf3eb83d76f27eddf21ca9db87c297612c07fad00807a601e03990feb5b6c9ad4"
  ... etc ...
})
```

Proposed docs:

```
>>> web3.eth.getBlock('latest')
AttributeDict({'number': 4023147, 'hash': "0xf3eb83d76f27eddf21ca9db87c297612c07fad00807a601e03990feb5b6c9ad4", ... etc ... })
```